adam/visualization/positioning.py

```diff
@@ -0,0 +1,426 @@
+    def create_at_random_position_scaled(
+        *,
+        min_distance_from_origin: float,
+        max_distance_from_origin: float,
+        object_scale: torch.Tensor,
+    ):
+        check_arg(min_distance_from_origin > 0.0)
+        check_arg(min_distance_from_origin < max_distance_from_origin)
+        # we first generate a random point on the unit sphere by
+        # generating a random vector in cube...
+        center = np.random.randn(3, 1).squeeze()
+        # and then normalizing.
+        center /= np.linalg.norm(center)
+
+        # then we scale according to the distances above
+        scale_factor = np.random.uniform(
+            min_distance_from_origin, max_distance_from_origin
+        )
+        center *= scale_factor
+        return AxisAlignedBoundingBox(
+            Parameter(
+                torch.tensor(center, dtype=torch.float),  # pylint: disable=not-callable
+                requires_grad=True,
+            ),
+            torch.diag(object_scale),
+        )
+
+    def get_corners(self) -> torch.Tensor:
+        return self.center.expand(8, 3) + torch.tensor(  # pylint: disable=not-callable
+            [
+                [-1, -1, -1],
+                [1, -1, -1],
+                [-1, 1, -1],
+                [1, 1, -1],
+                [-1, -1, 1],
+                [1, -1, 1],
+                [-1, 1, 1],
+                [1, 1, 1],
+            ],
+            dtype=torch.float,
+        ).matmul(self.scale)
+        # see https://github.com/pytorch/pytorch/issues/24807 re: pylint issue
+
+    # helper functions giving names to a few corners used in calculations:
+    def right_corner(self) -> torch.Tensor:
+        """
+        Corner in the direction of the positive x axis from center (negative direction from other axes).
+        (Assuming box is oriented to world axes)
+        Returns: Tensor (3,)
+        """
+        return self.center + torch.tensor(  # pylint: disable=not-callable
+            [1, -1, -1], dtype=torch.float
+        ).matmul(self.scale)
+        # see https://github.com/pytorch/pytorch/issues/24807 re: pylint issue
+
+    def forward_corner(self) -> torch.Tensor:
+        """
+        Corner in the direction of the positive y axis from center (negative direction from other axes).
+        (Assuming box is oriented to world axes)
+        Returns: Tensor (3,)
+
+        """
+        return self.center + torch.tensor(  # pylint: disable=not-callable
+            [-1, 1, -1], dtype=torch.float
+        ).matmul(self.scale)
+
+    def up_corner(self) -> torch.Tensor:
+        """
+        Corner in the direction of the positive z axis from center (negative direction from other axes).
+        (Assuming box is oriented to world axes)
+        Returns: Tensor (3,)
+
+        """
+        return self.center + torch.tensor(  # pylint: disable=not-callable
+            [-1, -1, 1], dtype=torch.float
+        ).matmul(self.scale)
+
+    def minus_ones_corner(self) -> torch.Tensor:
```

Comment:
This one is used in several places, but maybe prefix it by `_` since I am not sure we want it as part of the "public API" of the bounding box class.